tfjs-converter/src/executor/model_analysis.ts

```diff
@@ -101,96 +101,186 @@ export function getExecutionSubgraph(
-      frontier.push(input);
-    }
-  });
-  graph.weights.forEach(weight => {
-    if (usedNodes.has(weight.name)) {
-      frontier.push(weight);
-    }
-  });
-  if (initNodes != null) {
-    initNodes.forEach(node => {
-      if (usedNodes.has(node.name)) {
-        frontier.push(node);
+  const isUsed = (node: Node|string) =>
+      usedNodes.has(typeof node === 'string' ? node : node.name);
+
+  function unique(nodes: Node[]): Node[] {
+    return [...new Map(nodes.map((node) => [node.name, node])).values()];
+  }
+  const predefinedNodes = unique([
+                            ...inputNodes,
+                            ...graph.weights,
+                            ...initNodes,
+                          ]).filter(isUsed);
+  const allNodes = unique([
+                     ...predefinedNodes,
+                     ...Object.values(graph.nodes),
+                   ]).filter(isUsed);
+  const nameToNode =
+      new Map<string, Node>(allNodes.map((node) => [node.name, node]));
+
+  const inCounts: Record<string, number> = {};
+  for (const node of allNodes) {
+    inCounts[node.name] = inCounts[node.name] || 0;
+    for (const child of node.children) {
+      // When the child is unused, set in counts to infinity so that it will
+      // never be decreased to 0 and added to the execution list.
+      if (!isUsed(child)) {
+        inCounts[child.name] = Number.POSITIVE_INFINITY;
       }
-    });
+      inCounts[child.name] = (inCounts[child.name] || 0) + 1;
+    }
   }
-  const seen = new Set<string>();
-  const orderedNodes: Node[] = [];
+
+  // Build execution order for all used nodes regardless whether they are
+  // predefined or not.
+  const frontier = Object.entries(inCounts)
+                       .filter(([, inCount]) => inCount === 0)
+                       .map(([name]) => name);
+  const orderedNodeNames = [...frontier];
   while (frontier.length > 0) {
-    const node = frontier.pop();
-    seen.add(node.name);
-    if (!weightMap[node.name]) {
-      orderedNodes.push(node);
+    const nodeName = frontier.pop();
+    const node = nameToNode.get(nodeName)!;
+    for (const child of node.children.filter(isUsed)) {
+      if (--inCounts[child.name] === 0) {
+        orderedNodeNames.push(child.name);
+        frontier.push(child.name);
+      }
     }
-    node.children.forEach(child => {
-      if (!seen.has(child.name) && usedNodes.has(child.name) &&
-          child.inputs.every(input => seen.has(input.name))) {
-        frontier.push(child);
+  }
+
+  // Build a set for all nodes reachable by at least one predefined node.
+  // This can help us filter out redundant nodes from the returned node list.
+  // For example:
+  // If we have four nodes with dependencies like this:
+  //   a --> b --> c --> d
+  // when node `c` is predefined (e.g. given as an input tensor), we can
+  // skip node `a` and `b` since their outputs will never be used.
+  // TODO: Filter out more nodes when >=2 nodes are predefined in a path.
```

Comment:
This function intends to make the execution order, without changing the graph. And its' upstream mark unused nodes by providing a set of names in `usedNodes` instead of modifying the graph (removing nodes from children or inputs). So I don't want to break the contract, unless there is a plan to refactor everything starting from reading graph from a file.

At the time of making the execution plan, we don't know which nodes are output nodes. Therefore it just assumes all predefined reachable nodes to be output nodes. The output node list are provided by external caller in `GraphExecutor.execute` through:
1. Parameter `outputs?: string[]`
2. Env flag `KEEP_INTERMEDIATE_TENSORS`
Therefore, it is not something fixed with the graph. Of course we can make different execution plan for different set of "output node names", but I can see why the original author did not do that: imagine a user is debugging a weird bug by adding intermediate name to param outputs, which can make a different plan and cause the execution behavior to be different.

Besides, I don't think abstracting out graph logic (like topological order solver) is a good idea. It would end up with lots of redundant conversions between tfjs graph and the abstract graph, and make it harder to extend/modify for new tfjs graph specific properties and requirements. For now this function is self-contained for the purpose of making the execution plan (execution plan is not limited to topological order, so it may still better to rename this function with its purpose instead of it's implementation), and there are no other functions need the solver. So I think it's okay to keep it like this.

But I think wrapping it in a TFJSGraph class as a method will definitely be better. It will even better if we can separate SyncGraph and AsyncGraph with different checks/planner/executor. Go ahead if you have any good ideas to refactor this part!

Reply:
That makes sense. I'm happy to leave the graph as-is for this PR. We can talk more offline about how best to represent the graph in the future.